api/explorer/api/ingest.py

```diff
@@ -102,184 +72,64 @@ def _edge_from_build_input_type(build_input_type: model.BuildInputType) -> graph
-            for f in futures:
-                f.add_done_callback(lambda _x: bar.update(1))
-            wait(futures)
-            exceptions = [f.exception() for f in futures if f.exception()]
-            if exceptions:
-                # Raise a single exception after all tasks have completed
-                raise Exception(
-                    "Exceptions occurred while writing packages."
-                ) from exceptions[0]
-
-        click.echo("Writing edges")
-        with click.progressbar(length=len(edges)) as bar:
-            futures = [ex.submit(write_edge, e) for e in edges]
-            for f in futures:
-                f.add_done_callback(lambda _x: bar.update(1))
-            wait(futures)
-            for f in futures:
-                try:
-                    f.result()
-                except Exception as exc:
-                    logger.exception(exc)
-
-
-def ingest_nix_graph(
-    nix_graph: model.NixGraph, g_factory: Callable[[], GraphTraversalSource]
-) -> None:
-    """Writes the entire provided Nix graph to the provided Gremlin traversal source
-
-    Args:
-        nix_graph (model.NixGraph): The Nix graph to ingest
-        g (GraphTraversalSource): The graph traversal source to use for ingesting
-            the Nix graph
+    Writes the entire provided Nix graph to the provided Gremlin traversal source.
     """
-    ctxt = _IngestionContext()
-    for package in nix_graph.packages:
-        ingest_nix_package(package, ctxt, g_factory)
-
-    nb_missing_edges = sum(
-        len(src_edge_pairs) for src_edge_pairs in ctxt.pending_edges.values()
-    )
-    if nb_missing_edges > 0:
-        print(
-            f"Warning: {nb_missing_edges} build input(s) in the input JSON file "
-            "do not have a corresponding package and will be ignored."
-        )
+    for line in infile:
+        try:
+            derivation = model.Derivation.parse_raw(line)
+        except ValidationError as e:
+            click.echo(e, err=True)
+            continue
 
+        # convert from extraction to db model
+        # possibly handling multi-outputs derivation
+        derivation_nodes = core_to_graph(derivation)
+        build_input_nodes = [
+            graph.Derivation(
+                output_path=build_input.output_path,
+                attribute_path=None,
+            )
+            for build_input in derivation.build_inputs
+            if build_input.output_path is not None
+        ]
 
-@click.command(
-    help="Ingests a NixGraph JSON file to the specified Gremlin Server",
-    context_settings={"show_default": True},
-)
-@click.option(
-    "--graph-json",
+        # insert nodes
+        for node in derivation_nodes:
+            # insert if it does not exist, otherwise update properties
+            graph.upsert_unique_vertex(g, node)
+        for node in build_input_nodes:
+            # insert build input nodes to allow creating the edge
+            graph.insert_unique_vertex(g, node)
```

Comment:
We don't want to update from the data of `build_input_nodes` because it would update attributes like `attribute_path` with a less preferable value: we insert just to make sure there is a node to link the edge to, but this node will be later updated.